contexts/SecurityProvider.tsx

```diff
@@ -0,0 +1,305 @@
+        if (refreshWallet) {
+          refreshWallet();
+        }
+        if (lockAfterBackground === 'true') {
+          setLockedScreen(true);
+        }
+      }
+      appState.current = nextAppState;
+      setAppStateVisible(appState.current);
+    });
+
+    return () => {
+      subscription.remove();
+    };
+  }, [refreshWallet, lockAfterBackground, setLockedScreen]);
+
+  const {read} = useKeychain();
+
+  const [isPinOrFingerprintSet, setIsPinOrFingerprintSet] = useState<
+    boolean | undefined
+  >(undefined);
+
+  const [supportedBiometry, setSupportedBiometry] =
+    useState<Keychain.BIOMETRY_TYPE | null>(null);
+
+  useEffect((): void => {
+    DeviceInfo.isPinOrFingerprintSet().then(setIsPinOrFingerprintSet);
+    Keychain.getSupportedBiometryType({}).then(setSupportedBiometry);
+  }, []);
+
+  const [setManualPin, setSetManualPin] = useState<any>(() => {});
+  const [askManualPin, setAskManualPin] = useState<any>(() => {});
+
+  const [supportedType, setSupportedType] =
+    useState<SecurityAuthenticationTypes>(SecurityAuthenticationTypes.NONE);
+
+  const [error, setError] = useState<string | undefined>(undefined);
+
+  useEffect((): void => {
+    AsyncStorage.getItem('AuthenticationType').then(async val => {
+      if (!val) {
+        if (supportedBiometry !== null) {
+          AsyncStorage.setItem(
+            'AuthenticationType',
+            SecurityAuthenticationTypes.KEYCHAIN,
+          );
+          setSupportedType(SecurityAuthenticationTypes.KEYCHAIN);
+          setError(undefined);
+        } else if (isPinOrFingerprintSet === true) {
+          AsyncStorage.setItem(
+            'AuthenticationType',
+            SecurityAuthenticationTypes.LOCALAUTH,
+          );
+          setSupportedType(SecurityAuthenticationTypes.LOCALAUTH);
+          setError(undefined);
+        } else {
+          AsyncStorage.setItem(
+            'AuthenticationType',
+            SecurityAuthenticationTypes.MANUAL,
+          );
+          setSupportedType(SecurityAuthenticationTypes.MANUAL);
+          setError(undefined);
+        }
+      } else {
+        if (
+          supportedBiometry === null &&
+          val == SecurityAuthenticationTypes.KEYCHAIN
+        ) {
+          // Keychain not available anymore
+          setError(
+            'Your previous authentication method is not available anymore. Please enroll your biometrics on your device again or reinstall the app.',
+          );
+        } else if (
+          isPinOrFingerprintSet !== true &&
+          val == SecurityAuthenticationTypes.LOCALAUTH
+        ) {
+          // LocalAuth not available anymore
+          setError(
+            'Your previous authentication method is not available anymore. Please add a PIN code lock to your device or reinstall the app.',
+          );
+        } else {
+          setSupportedType(val as SecurityAuthenticationTypes);
+          setError(undefined);
+        }
```

Comment:
I think it wud be better to add a little bit documentation for each if case such that we can easily understand when we come back to here. It was quite confused for me and took me a while to understand. (maybe i dont have enought context prevly)

Reply:
I will add some docs on the file